client-apis/streaming.md

```diff
@@ -0,0 +1,112 @@
+
+# Streaming
+
+## Version
+
+0.1.0
+
+## Changelog
+
+- 0.1.0
+
+  - [Clarify Streaming Encrypt and Decrypt](../changes/2020-07-06_clarify-streaming-encrypt-decrypt/change.md)
+
+## Overview
+
+The AWS Encryption SDK MAY provide APIs that enable streamed [encryption](encrypt.md)
+and [decryption](decrypt.md).
+Streaming is a framework for making bytes available to be processed
+by an operation sequentially and over time,
+and for outputting the result of that processing
+sequentially and over time.
+
+If an implementation requires holding the entire input in memory in order to perform the operation,
+that implementation SHOULD NOT provide an API that allows the caller to stream the operation.
+APIs that support streaming of the encrypt or decrypt operation SHOULD allow customers
+to be able to process arbitrarily large inputs with a finite amount of working memory.
+
+## Definitions
+
+### Conventions used in this document
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in [RFC2119](https://tools.ietf.org/html/rfc2119).
+
+### Available Bytes
+
+In the scope of an operation, bytes are considered consumable if:
+
+- The operation has not yet processed those bytes,
+- The operation has access to those bytes,
+- Those bytes are intended to be processed.
+  This intention is expressed through the specific streaming interface.
+
+For example, in a framework where a customer is sending input bytes to an operation
+and that operation must write those bytes to some sink,
+the input bytes received from the customer are considered consumable.
+Here the customer is expressing intent to process their supplied bytes.
+
+For a framework where a customer is requesting output bytes from an operation
+and that operation must read from some source in order to produce bytes,
+this is slightly more complicated.
+Bytes are considered consumable if:
+
+- Those bytes have not yet been processed.
+- Those bytes are able to be read by the operation from the source.
+- Those bytes are required to be processed in order for the operation
+  to release the output requested by the customer.
+  Here the customer expresses intent for the operation to process
+  whatever the operation needs to consume to produce its complete output
+
+### Release
+
+An operation releases processed bytes when those bytes are outputted by the operation.
+
+If bytes are processed by an operation, that does not imply that the operation is allowed to
+release those bytes.
+The decrypt and encrypt operations specify when processed bytes MAY be released,
+when they SHOULD be released,
+and when they MUST NOT be released.
+
+## Inputs
+
+In order to support streaming, the operation MUST accept some input within a streaming framework.
+
+This means that:
+
+- There MUST be a mechanism for input bytes to become consumable.
```

Comment:
Perhaps imply a little more strongly that this is referring to a specific sequence of bytes at a time, and happens repeatedly? Otherwise it reads a bit as "a mechanism for bytes to flow in some way over time"